ironfish/src/rpc/adapters/socketAdapter/socketAdapter.ts

```diff
@@ -0,0 +1,281 @@
+        },
+      )
+    })
+  }
+
+  async stop(): Promise<void> {
+    if (!this.started) {
+      return
+    }
+
+    this.clients.forEach((client) => {
+      client.requests.forEach((r) => r.close())
+      client.socket.destroy()
+    })
+
+    await new Promise<void>((resolve, reject) => {
+      this.server?.close((error) => {
+        if (error) {
+          reject(error)
+        } else {
+          resolve()
+        }
+      })
+    })
+
+    await this.waitForAllToDisconnect()
+
+    this.logger.debug(`SocketAdapter stopped: ${this.host}:${this.port}`)
+  }
+
+  attach(server: RpcServer): void {
+    this.router = server.getRouter(this.namespaces)
+  }
+
+  async waitForAllToDisconnect(): Promise<void> {
+    const clients = Array.from(this.clients.values())
+    await Promise.all(clients.map((c) => this.waitForClientToDisconnect(c)))
+  }
+
+  waitForClientToDisconnect(client: SocketClient): Promise<void> {
+    return new Promise<void>((resolve) => {
+      client.socket.once('close', () => {
+        resolve()
+      })
+    })
+  }
+
+  onClientConnection(socket: net.Socket): void {
+    const requests = new Map<string, Request>()
+    const client = { socket, requests, id: uuid() }
+    this.clients.set(client.id, client)
+
+    socket.on('data', (data) => {
+      this.onClientData(client, data).catch((e) => {
+        this.onClientError(client, e)
+      })
+    })
+
+    socket.on('close', () => {
+      this.onClientDisconnection(client)
+    })
+
+    socket.on('error', (error: Error) => {
+      this.onClientError(client, error)
+    })
+  }
+
+  onClientDisconnection(client: SocketClient): void {
+    client.requests.forEach((req) => req.close())
+    this.clients.delete(client.id)
+    this.logger.debug(`client connection closed: ${this.host}:${this.port}`)
+  }
+
+  onClientError(client: SocketClient, error: unknown): void {
+    this.logger.debug(`${this.host}:${this.port} has error: ${ErrorUtils.renderError(error)}`)
+  }
+
+  async onClientData(client: SocketClient, data: Buffer): Promise<void> {
+    const [parsed, error] = JSONUtils.tryParse(data.toString('utf8').trim())
+    if (error) {
+      this.emitResponse(client, this.constructUnmountedAdapter())
+      return
+    }
+
+    const result = await YupUtils.tryValidate(ClientSocketRpcSchema, parsed)
+
+    if (result.error) {
+      this.emitResponse(client, this.constructMalformedRequest(parsed))
+      return
+    }
+
+    const message = result.result.data
+
+    const requestId = uuid()
+    const request = new Request(
+      message.data,
+      (status: number, data?: unknown) => {
+        this.emitResponse(client, this.constructMessage(message.mid, status, data), requestId)
+      },
+      (data: unknown) => {
+        this.emitStream(client, this.constructStream(message.mid, data))
+      },
+    )
+    client.requests.set(requestId, request)
+
+    if (this.router == null) {
+      this.emitResponse(client, this.constructMalformedRequest(data))
```

Comment:
This one seems like an unmounted adapter rather than a malformed request

Reply:
Changed